tools/state-viewer/src/state_parts.rs

```diff
@@ -0,0 +1,432 @@
+        near_config.genesis.config.genesis_height,
+        near_config.client_config.save_trie_changes,
+    );
+
+    let epoch_id = epoch_selection.to_epoch_id(store, &mut chain_store, &mut epoch_manager);
+    let epoch = runtime_adapter.get_epoch_info(&epoch_id).unwrap();
+    let sync_prev_hash = get_prev_hash_of_epoch(&epoch, &mut chain_store, &mut epoch_manager);
+    let sync_prev_block = chain_store.get_block(&sync_prev_hash).unwrap();
+
+    assert!(runtime_adapter.is_next_block_epoch_start(&sync_prev_hash).unwrap());
+    assert!(
+        shard_id < sync_prev_block.chunks().len() as u64,
+        "shard_id: {}, #shards: {}",
+        shard_id,
+        sync_prev_block.chunks().len()
+    );
+    let state_root = sync_prev_block.chunks()[shard_id as usize].prev_state_root();
+
+    let part_storage: Box<dyn StatePartReader> = if let Some(root_dir) = root_dir {
+        if s3_bucket.is_some() || s3_region.is_some() {
+            panic!("Output can only be done to the local filesystem or to S3, not both");
+        }
+        Box::new(FileSystemStorage::new(
+            root_dir,
+            false,
+            &near_config.client_config.chain_id,
+            epoch.epoch_height(),
+            shard_id,
+        ))
+    } else {
+        if s3_bucket.is_none() || s3_region.is_none() {
+            panic!("For storing parts on S3 both --s3-bucket and --s3-region need to be provided");
+        }
+        Box::new(S3Storage::new(
+            &s3_bucket.unwrap(),
+            &s3_region.unwrap(),
+            &near_config.client_config.chain_id,
+            epoch.epoch_height(),
+            shard_id,
+        ))
+    };
+
+    let num_parts = part_storage.num_parts();
+    tracing::info!(
+        target: "state-parts",
+        epoch_height = epoch.epoch_height(),
+        epoch_id = ?epoch_id.0,
+        shard_id,
+        num_parts,
+        ?sync_prev_hash,
+        part_id = if let Some(part_id) = part_id { part_id.to_string() } else { "<all_parts>".to_string() },
+        "Applying state as seen at the beginning of the specified epoch.",
+    );
+
+    let timer = Instant::now();
+    for part_id in if let Some(part_id) = part_id { part_id..part_id + 1 } else { 0..num_parts } {
+        let timer = Instant::now();
+        assert!(part_id < num_parts, "part_id: {}, num_parts: {}", part_id, num_parts);
+        let part = part_storage.read(part_id);
+        runtime_adapter
+            .apply_state_part(
+                shard_id,
+                &state_root,
+                PartId::new(part_id, num_parts),
+                &part,
+                &epoch_id,
+            )
+            .unwrap();
+        tracing::info!(target: "state-parts", part_id, part_length = part.len(), elapsed_sec = timer.elapsed().as_secs_f64(), "Applied a state part");
+    }
+    tracing::info!(target: "state-parts", total_elapsed_sec = timer.elapsed().as_secs_f64(), "Applied all requested state parts");
+}
+
+pub(crate) fn dump_state_parts(
```

Comment:
There seem to be some similarities between the apply and dump function, can you refactor it to common helper functions?

Reply:
The loops need that common code to get 6 objects. Extracting it to a common function is possible, but seems too rigid because it's too specific.
Extracting the two different loops into closures, also seems too rigid.

Reply:
Moved some common code into separate tiny functions.